src/details/ArborX_DetailsDendrogram.hpp

```diff
@@ -0,0 +1,170 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_DETAILS_DENDROGRAM_HPP
+#define ARBORX_DETAILS_DENDROGRAM_HPP
+
+#include <ArborX_DetailsKokkosExtSwap.hpp>
+#include <ArborX_DetailsKokkosExtViewHelpers.hpp>
+#include <ArborX_DetailsSortUtils.hpp>
+#include <ArborX_DetailsUnionFind.hpp>
+#include <ArborX_MinimumSpanningTree.hpp> // WeightedEdge
+
+#include <Kokkos_Core.hpp>
+#include <Kokkos_Profiling_ProfileSection.hpp>
+
+namespace ArborX::Details
+{
+
+struct UnweightedEdge
+{
+  int source;
+  int target;
+};
+
+// Compute permutation that orders edges in increasing order
+template <typename ExecutionSpace, typename MemorySpace>
+Kokkos::View<unsigned int *, MemorySpace>
+computeEdgesPermutation(ExecutionSpace const &exec_space,
+                        Kokkos::View<WeightedEdge *, MemorySpace> &edges)
+{
+  Kokkos::Profiling::pushRegion("ArborX::Dendrogram::sort_edges");
+
+  int const num_edges = edges.size();
+
+  Kokkos::View<float *, MemorySpace> weights(
+      Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                         "ArborX::Dendrogram::weights"),
+      num_edges);
+  Kokkos::parallel_for(
+      "ArborX::Dendrogram::copy_weights",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_edges),
+      KOKKOS_LAMBDA(int const e) { weights(e) = edges(e).weight; });
+
+  auto permute = Details::sortObjects(exec_space, weights);
+
+  Kokkos::Profiling::popRegion();
+
+  return permute;
+}
+
+template <typename Edges, typename Parents>
+void dendrogramUnionFindHost(Edges sorted_edges_host, Parents &parents_host)
```

Comment:
I find it confusing that you call the function doing the actual work `*Host`. Maybe, rather
```suggestion
void dendrogramUnionFindImpl(Edges sorted_edges_host, Parents &parents_host)
```
? Otherwise, this suggests to me that there is a host and a device version.

Reply:
I added `Host` to make sure it is clear that the main kernel will be executed on the host.

Reply:
Yes, and I find it confusing.